gateway/test/src/test/java/io/apiman/gateway/test/junit/vertx3/Vertx3GatewayFileRegistryServer.java

```diff
@@ -0,0 +1,162 @@
+     */
+    public Vertx3GatewayFileRegistryServer() {
+    }
+
+    @Override
+    public void configure(JsonNode nodeConfig) {
+        vertxConf = loadJsonObjectFromResources(nodeConfig, "config");
+        pushEmulatorConfig = loadJsonObjectFromResources(nodeConfig, "configPushEmulator");
+
+        secondVx = Vertx.vertx(new VertxOptions()
+                .setBlockedThreadCheckInterval(99999));
+
+        secondVx.deployVerticle(ApiVerticle.class.getCanonicalName(),
+                new DeploymentOptions().setConfig(pushEmulatorConfig));
+    }
+
+    private JsonObject loadJsonObjectFromResources(JsonNode nodeConfig, String name) {
+        ClassLoader classLoader = getClass().getClassLoader();
+        String fPath = nodeConfig.get(name).asText();
+        File file = new File(classLoader.getResource(fPath).getFile());
+        try {
+            conf = new String(Files.readAllBytes(file.toPath()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return new JsonObject(conf);
+    }
+
+    @Override
+    public String getApiEndpoint() {
+        return "http://localhost:" + API_PORT;
+    }
+
+    @Override
+    public String getGatewayEndpoint() {
+        return "http://localhost:" + GW_PORT;
+    }
+
+    @Override
+    public String getEchoTestEndpoint() {
+        return "http://localhost:" + ECHO_PORT;
+    }
+
+    @Override
+    public void start() {
+        try {
+            vertx = Vertx.vertx(new VertxOptions()
+                    .setBlockedThreadCheckInterval(99999));
+            echoServer.start();
+
+            startLatch = new CountDownLatch(1);
+            DeploymentOptions options = new DeploymentOptions();
+            options.setConfig(vertxConf);
+
+            vertx.deployVerticle(InitVerticle.class.getCanonicalName(),
+                    options, event -> {
+                        System.out.println("Deployed init verticle!");
+                        startLatch.countDown();
+                    });
+
+            startLatch.await();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void stop() {
+        System.err.println("Stopping main Vert.x");
+        stopLatch = new CountDownLatch(1);
+        echoServer.stop();
+
+        secondVx.eventBus().publish(ApiToFileRegistry.class.getCanonicalName(), null,
+                new DeliveryOptions().addHeader("action", "reset"));
+        URILoadingRegistry.reset();
+
+        vertx.close(result -> {
+            stopLatch.countDown();
+        });
+
+        try {
+            stopLatch.await();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void next(String endpoint) {
+        try {
+            if (URI.create(endpoint).getPort() == API_PORT) {
+                CountDownLatch cdl = new CountDownLatch(1);
+                URILoadingRegistry.reloadData(done -> {
```

Comment:
Had to add some extra functionality to the registry to force to reload. This is likely to be useful anyway as a true polling/pull registry rather than a one-shot immutable pull registry.